src/components/AssetInput/AssetInput.tsx

```diff
@@ -0,0 +1,230 @@
+import IconCaretDown from "@elements/Icon/Generated/IconCaretDown";
+import IconImageLibrary from "@elements/Icon/Generated/IconImageLibrary";
+import IconUploadAlternative from "@elements/Icon/Generated/IconUploadAlternative";
+import { IconProps } from "@elements/Icon/IconProps";
+import { IconSize } from "@elements/Icon/IconSize";
+import { useButton } from "@react-aria/button";
+import { FocusScope, useFocusRing } from "@react-aria/focus";
+import { useMenuTrigger } from "@react-aria/menu";
+import { DismissButton, useOverlay } from "@react-aria/overlays";
+import { useMenuTriggerState } from "@react-stately/menu";
+import { FOCUS_STYLE } from "@utilities/focusStyle";
+import { merge } from "@utilities/merge";
+import { AnimatePresence, motion } from "framer-motion";
+import React, { cloneElement, FC, ReactElement, useRef } from "react";
+
+type OtherAsset = {
+    name: string;
+    type: "audio";
+    extension: string;
+    src?: undefined;
+    icon?: undefined;
+    size: number;
+};
+
+type ImageAsset = {
+    name: string;
+    type: "image" | "logo";
+    extension: string;
+    src: string;
+    icon?: undefined;
+    size: number;
+};
+
+type IconAsset = {
+    name: string;
+    type: "icon";
+    extension?: undefined;
+    src?: undefined;
+    icon: ReactElement<IconProps>;
+    size?: undefined;
+};
+
+type AssetProps = {
+    asset:
+        | (ImageAsset & { source: "upload" })
+        | (ImageAsset & { source: "library"; sourceName: string })
+        | (IconAsset & { source: "upload" })
+        | (IconAsset & { source: "library"; sourceName: string })
+        | (OtherAsset & { source: "upload" })
+        | (OtherAsset & { source: "library"; sourceName: string });
+    size: "small" | "large";
+    actions: ActionMenuProps["menuBlocks"];
+};
+
+export type AssetInputProps =
+    | (AssetProps & {
+          onUploadClick?: undefined;
+          onLibraryClick?: undefined;
+      })
+    | {
+          asset?: undefined;
+          actions?: undefined;
+          size?: undefined;
+          onUploadClick: () => void;
+          onLibraryClick?: () => void;
+      };
+
+const SelectedAsset: FC<AssetProps> = ({ asset, size, actions }) => {
+    const menu = useMenuTriggerState({ closeOnSelect: true });
+    const buttonRef = useRef<HTMLButtonElement | null>(null);
+    const { menuTriggerProps } = useMenuTrigger({}, menu, buttonRef);
+    const { buttonProps } = useButton(menuTriggerProps, buttonRef);
+    const { isFocusVisible, focusProps } = useFocusRing();
+    const overlayRef = useRef<HTMLDivElement | null>(null);
+    const { overlayProps } = useOverlay(
+        { onClose: () => menu.close(), shouldCloseOnBlur: true, isOpen: menu.isOpen, isDismissable: true },
+        overlayRef,
+    );
+    const { isOpen } = menu;
+
+    return (
+        <div className="tw-relative tw-w-full tw-font-sans tw-text-s">
+            <button
+                {...buttonProps}
+                {...focusProps}
```

Comment:
React-aria provides a `mergeProps` method, maybe useful in this case: https://react-spectrum.adobe.com/react-aria/mergeProps.html